Src/IronPython/Modules/Builtin.cs

```diff
@@ -325,26 +325,45 @@ public static partial class Builtin {
         }
 
         [LightThrowing]
-        public static object eval(CodeContext/*!*/ context, string expression) => eval(context, expression, null, null);
+        public static object eval(CodeContext/*!*/ context, object expression) => eval(context, expression, null, null);
 
         [LightThrowing]
-        public static object eval(CodeContext/*!*/ context, string expression, PythonDictionary globals) => eval(context, expression, globals, null);
+        public static object eval(CodeContext/*!*/ context, object expression, PythonDictionary globals) => eval(context, expression, globals, null);
 
         [LightThrowing]
-        public static object eval(CodeContext/*!*/ context, string expression, PythonDictionary globals, object locals) {
+        public static object eval(CodeContext/*!*/ context, object expression, PythonDictionary globals, object locals) {
             Debug.Assert(context != null);
-            if (expression == null) throw PythonOps.TypeError("eval() argument 1 must be string or code object");
+
+            string strExpression;
+            switch (expression) {
+                case string s:
+                    strExpression = s;
+                    break;
+
+                case ByteArray ba:
+                    strExpression = ba.ToString();
```

Comment:
Yes, `ToString` is ultimately not right, but neither is `StringOps.RawDecode` decoding to UTF-8, although the latter is better in the Ptyhon 3 context so I will use it instead of `ToString`. 

My intention was to do minimal changes to be able to run `test_source_encoding` (and fail, rather than bomb with errors as now). Then tackle Issue #22, which @slide  believed is a "good first issue". IronPython is a steep entry curve for me and I don't want to break anything by lack of understanding. When Issue #22 is resolved, `test_source_encoding` should pass and be enabled in the standard set of tests.

A side note on `Bytes.ToString()`: as you say it blindly casts bytes to chars, but more specifically, it unconditionally decodes bytes as Latin 1. This may have been a good choice for Python 2, which assumes Latin 1 as the default encoding, but Python 3 assumes UTF-8 as default. However, `ToString` is a .NET convention, not Python. Anyway, I don't think that assuming Latin 1 is right here. I don't know what IronPython uses `ToString` for (except for the three cases discussed in this PR, which I will eliminate anyway), hopefully for nothing, in which case it is useful only for debugging and client code. In such case I think it is better to represent all 8-bit-set characters and all non-printable characters in `\xXX` form and enclose the string in `b'...'`, which is what Python 3 would print out:
```python
print(b"\xd0\x9f\xd0\xb8\xd1\x82\xd0\xbe\xd0\xbd")
```
Python 3.4: `b'\xd0\x9f\xd0\xb8\xd1\x82\xd0\xbe\xd0\xbd'`
IronPython 3: `ÐÐ¸ÑÐ¾Ð½`

Reply:
I think we're already reading source files using UTF-8 (#22) at the moment, however, we're not handling the error cases properly. Probably needs to be something like:
```C#
enc = (Encoding)Encoding.UTF8.Clone();
enc.DecoderFallback = new SourceNonStrictDecoderFallback();
```
in
https://github.com/IronLanguages/ironpython3/blob/b3fdd951a150be1b2ee376576e260bbb218478d7/Src/IronPython/Runtime/PythonContext.cs#L939

Unfortunately I don't think my sample will work as-is since UTF8 is bugged. Note that checking the encoding in headers is also done in there.

Reply:
What do you mean with "UTF8 is bugged"?

Reply:
Using `DecoderFallback` with `Encoding.UTF8` has a bug so that if you try to use `SourceNonStrictDecoderFallback` with it, it may throw an index out of range exception.